File: backend/src/modules/jobs/job.service.ts
import { randomUUID } from "node:crypto";
import type { PoolClient } from "pg";

import { AppError } from "../../shared/app-error.js";
import { JobRepository } from "./job.repository.js";
import type {
  JobDraftInput,
  JobDraftUpdateInput,
  JobStatus,
  RequestMetadata,
  ReviewDecision,
} from "./job.types.js";

type Actor = { userId: string; companyId: string | null };

function notFound() {
  return new AppError(404, "JOB_NOT_FOUND", "Không tìm thấy tin tuyển dụng.");
}

function conflict(message: string) {
  return new AppError(409, "JOB_STATE_CONFLICT", message);
}

function isExpired(deadline: string | Date) {
  const value = typeof deadline === "string" ? deadline.slice(0, 10) : deadline.toISOString().slice(0, 10);
  const today = new Date().toISOString().slice(0, 10);
  return value < today;
}

export class JobService {
  constructor(private readonly repository: JobRepository) {}

  async listCompanyJobs(companyId: string | null, input: { page: number; pageSize: number; status?: JobStatus }) {
    if (!companyId) throw notFound();
    return this.repository.listCompanyJobs(companyId, input);
  }

  async getCompanyJob(companyId: string | null, jobId: string) {
    if (!companyId) throw notFound();
    const job = await this.repository.findById(jobId);
    if (!job || job.company.id !== companyId) throw notFound();
    return job;
  }

  async listReviewQueue(input: { page: number; pageSize: number }) {
    return this.repository.listReviewQueue(input);
  }

  async createDraft(actor: Actor, input: JobDraftInput, request: RequestMetadata) {
    if (!actor.companyId) throw notFound();
    return this.repository.withTransaction(async (client) => {
      const company = await client.query<{ partner_status: string }>(
        "SELECT partner_status FROM companies WHERE id = $1 FOR SHARE",
        [actor.companyId],
      );
      if (!company.rows[0]) throw notFound();
      if (company.rows[0].partner_status !== "ACTIVE") {
        throw new AppError(403, "COMPANY_NOT_ACTIVE", "Doanh nghiệp chưa ở trạng thái đối tác hoạt động.");
      }
      await this.validateReferences(client, input);
      const result = await client.query<{ id: string }>(
        `INSERT INTO job_posts
         (company_id, created_by_user_id, title, opportunity_type, work_mode, location,
          description, requirements, benefits, positions, deadline, status)
         VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, 'DRAFT')
         RETURNING id`,
        [
          actor.companyId,
          actor.userId,
          input.title,
          input.opportunityType,
          input.workMode,
          input.location,
          input.description,
          input.requirements,
          input.benefits,
          input.positions,
          input.deadline,
        ],
      );
      const jobId = result.rows[0]!.id;
      await this.repository.replaceReferences(client, jobId, input);
      await client.query(
        `INSERT INTO job_post_status_history
         (job_post_id, command_id, from_status, to_status, actor_type, actor_user_id)
         VALUES ($1, $2, NULL, 'DRAFT', 'COMPANY', $3)`,
        [jobId, randomUUID(), actor.userId],
      );
      await this.repository.writeAudit(client, {
        actorUserId: actor.userId,
        action: "JOB_DRAFT_CREATED",
        jobId,
        request,
      });
      return (await this.repository.findById(jobId, client))!;
    });
  }

  async updateDraft(
    actor: Actor,
    jobId: string,
    input: JobDraftUpdateInput,
    request: RequestMetadata,
  ) {
    if (!actor.companyId) throw notFound();
    return this.repository.withTransaction(async (client) => {
      const locked = await this.repository.lockJob(client, jobId);
      if (!locked || locked.companyId !== actor.companyId) throw notFound();
      if (!(["DRAFT", "REVISION_REQUIRED"] as JobStatus[]).includes(locked.status)) {
        throw conflict("Chỉ có thể sửa tin ở trạng thái Bản nháp hoặc Cần chỉnh sửa.");
      }
      if (locked.version !== input.expectedVersion) {
        throw conflict("Tin đã được cập nhật ở nơi khác. Vui lòng tải lại dữ liệu.");
      }
      await this.validateReferences(client, input);
      await client.query(
        `UPDATE job_posts SET
           title = $2, opportunity_type = $3, work_mode = $4, location = $5,
           description = $6, requirements = $7, benefits = $8, positions = $9,
           deadline = $10, version = version + 1
         WHERE id = $1`,
        [
          jobId,
          input.title,
          input.opportunityType,
          input.workMode,
          input.location,
          input.description,
          input.requirements,
          input.benefits,
          input.positions,
          input.deadline,
        ],
      );
      await this.repository.replaceReferences(client, jobId, input);
      await this.repository.writeAudit(client, {
        actorUserId: actor.userId,
        action: "JOB_DRAFT_UPDATED",
        jobId,
        metadata: { fromVersion: locked.version },
        request,
      });
      return (await this.repository.findById(jobId, client))!;
    });
  }

  async submit(actor: Actor, jobId: string, commandId: string, request: RequestMetadata) {
    if (!actor.companyId) throw notFound();
    return this.repository.withTransaction(async (client) => {
      if (await this.repository.commandExists(client, jobId, commandId)) {
        const repeated = await this.repository.findById(jobId, client);
        if (!repeated || repeated.company.id !== actor.companyId) throw notFound();
        return repeated;
      }
      const locked = await this.repository.lockJob(client, jobId);
      if (!locked || locked.companyId !== actor.companyId) throw notFound();
      if (await this.repository.commandExists(client, jobId, commandId)) {
        return (await this.repository.findById(jobId, client))!;
      }
      if (!(["DRAFT", "REVISION_REQUIRED"] as JobStatus[]).includes(locked.status)) {
        throw conflict("Tin không còn ở trạng thái có thể gửi UIT duyệt.");
      }
      if (locked.partnerStatus !== "ACTIVE") {
        throw new AppError(403, "COMPANY_NOT_ACTIVE", "Doanh nghiệp chưa ở trạng thái đối tác hoạt động.");
      }
      if (isExpired(locked.deadline)) {
        throw new AppError(400, "JOB_DEADLINE_EXPIRED", "Hạn ứng tuyển phải từ hôm nay trở đi.");
      }
      await client.query(
        `UPDATE job_posts SET status = 'PENDING_UIT_REVIEW', version = version + 1,
          submitted_at = now(), reviewed_at = NULL, reviewed_by_user_id = NULL WHERE id = $1`,
        [jobId],
      );
      await client.query(
        `INSERT INTO job_post_status_history
         (job_post_id, command_id, from_status, to_status, actor_type, actor_user_id)
         VALUES ($1, $2, $3, 'PENDING_UIT_REVIEW', 'COMPANY', $4)`,
        [jobId, commandId, locked.status, actor.userId],
      );
      const job = (await this.repository.findById(jobId, client))!;
      await this.notifyUitAdmins(client, jobId, commandId, job.title, job.company.name);
      await this.repository.writeAudit(client, {
        actorUserId: actor.userId,
        action: "JOB_SUBMITTED_FOR_UIT_REVIEW",
        jobId,
        metadata: { fromStatus: locked.status, commandId },
        request,
      });
      return job;
    });
  }

  async review(
    actorUserId: string,
    jobId: string,
    commandId: string,
    decision: ReviewDecision,
    reason: { reasonCode: string; note: string } | null,
    request: RequestMetadata,
  ) {
    return this.repository.withTransaction(async (client) => {
      if (await this.repository.commandExists(client, jobId, commandId)) {
        const repeated = await this.repository.findById(jobId, client);
        if (!repeated) throw notFound();
        return repeated;
      }
      const locked = await this.repository.lockJob(client, jobId);
      if (!locked) throw notFound();
      if (await this.repository.commandExists(client, jobId, commandId)) {
        return (await this.repository.findById(jobId, client))!;
      }
      if (locked.status !== "PENDING_UIT_REVIEW") {
        throw conflict("Tin đã được xử lý hoặc không còn chờ UIT duyệt.");
      }
      if (decision === "approve" && isExpired(locked.deadline)) {
        throw new AppError(400, "JOB_DEADLINE_EXPIRED", "Không thể duyệt tin đã hết hạn ứng tuyển.");
      }
      const nextStatus: JobStatus =
        decision === "approve" ? "RECRUITING" : decision === "request-revision" ? "REVISION_REQUIRED" : "REJECTED";
      await client.query(
        `UPDATE job_posts SET status = $2, version = version + 1,
          reviewed_by_user_id = $3, reviewed_at = now() WHERE id = $1`,
        [jobId, nextStatus, actorUserId],
      );
      await client.query(
        `INSERT INTO job_post_status_history
         (job_post_id, command_id, from_status, to_status, actor_type, actor_user_id, reason_code, note)
         VALUES ($1, $2, 'PENDING_UIT_REVIEW', $3, 'UIT_ADMIN', $4, $5, $6)`,
        [jobId, commandId, nextStatus, actorUserId, reason?.reasonCode ?? null, reason?.note ?? null],
      );
      const job = (await this.repository.findById(jobId, client))!;
      await this.notifyCompany(client, job.company.id, jobId, commandId, job.title, decision, reason?.note);
      await this.repository.writeAudit(client, {
        actorUserId,
        action: `JOB_UIT_${decision.replace("-", "_").toUpperCase()}`,
        jobId,
        metadata: { commandId, nextStatus, reasonCode: reason?.reasonCode ?? null },
        request,
      });
      return job;
    });
  }

  private async validateReferences(client: PoolClient, input: JobDraftInput) {
    const [categoriesValid, skillsValid] = await Promise.all([
      this.repository.assertReferenceIds(client, "categories", input.categoryIds),
      this.repository.assertReferenceIds(client, "skills", input.skillIds),
    ]);
    if (!categoriesValid || !skillsValid) {
      throw new AppError(400, "JOB_REFERENCE_INVALID", "Nhóm ngành hoặc kỹ năng không hợp lệ.");
    }
  }

  private async notifyUitAdmins(
    client: PoolClient,
    jobId: string,
    commandId: string,
    title: string,
    companyName: string,
  ) {
    await client.query(
      `INSERT INTO notifications
       (recipient_user_id, type, title, body, resource_type, resource_id, deep_link, dedupe_key, payload)
       SELECT u.id, 'JOB_PENDING_UIT_REVIEW', 'Có tin tuyển dụng mới cần duyệt',
              $3 || ' đã gửi tin "' || $4 || '".', 'JOB_POST', $1::uuid,
              '/uit/jobs/review', 'job:' || $1::text || ':submit:' || $2::text || ':' || u.id::text,
              jsonb_build_object('jobId', $1::text, 'commandId', $2::text)
       FROM users u WHERE u.role = 'UIT_ADMIN' AND u.status = 'ACTIVE'
       ON CONFLICT (dedupe_key) DO NOTHING`,
      [jobId, commandId, companyName, title],
    );
  }

  private async notifyCompany(
    client: PoolClient,
    companyId: string,
    jobId: string,
    commandId: string,
    title: string,
    decision: ReviewDecision,
    note?: string,
  ) {
    const copy = {
      approve: ["JOB_APPROVED", "Tin tuyển dụng đã được duyệt", `Tin "${title}" đã được công khai.`],
      "request-revision": ["JOB_REVISION_REQUIRED", "Tin tuyển dụng cần chỉnh sửa", note ?? `Tin "${title}" cần chỉnh sửa.`],
      reject: ["JOB_REJECTED", "Tin tuyển dụng đã bị từ chối", note ?? `Tin "${title}" đã bị từ chối.`],
    }[decision];
    await client.query(
      `INSERT INTO notifications
       (recipient_user_id, type, title, body, resource_type, resource_id, deep_link, dedupe_key, payload)
       SELECT u.id, $4, $5, $6, 'JOB_POST', $1::uuid, '/company/jobs/' || $1::text,
              'job:' || $1::text || ':review:' || $2::text || ':' || u.id::text,
              jsonb_build_object('jobId', $1::text, 'commandId', $2::text, 'decision', $7::text)
       FROM company_users cu JOIN users u ON u.id = cu.user_id
       WHERE cu.company_id = $3 AND u.status = 'ACTIVE'
       ON CONFLICT (dedupe_key) DO NOTHING`,
      [jobId, commandId, companyId, copy[0], copy[1], copy[2], decision],
    );
  }
}

File: backend/src/main/java/vn/edu/uit/careerhub/jobs/JobService.java
package vn.edu.uit.careerhub.jobs;

import java.time.LocalDate;
import java.time.ZoneOffset;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.UUID;

import org.springframework.http.HttpStatus;
import org.springframework.jdbc.core.simple.JdbcClient;
import org.springframework.stereotype.Service;
import org.springframework.transaction.support.TransactionTemplate;

import vn.edu.uit.careerhub.auth.AuthPrincipal;
import vn.edu.uit.careerhub.auth.AuthRepository;
import vn.edu.uit.careerhub.auth.RequestMetadata;
import vn.edu.uit.careerhub.common.AppException;
import vn.edu.uit.careerhub.jobs.JobModels.LockedJob;
import vn.edu.uit.careerhub.jobs.JobModels.ReviewDecision;
import vn.edu.uit.careerhub.jobs.JobModels.Status;
import vn.edu.uit.careerhub.jobs.JobRequests.Draft;
import vn.edu.uit.careerhub.jobs.JobRequests.ReviewReason;

@Service
public class JobService {
    private final JobRepository repository;
    private final AuthRepository audit;
    private final JdbcClient database;
    private final TransactionTemplate transactions;

    public JobService(JobRepository repository, AuthRepository audit, JdbcClient database, TransactionTemplate transactions) {
        this.repository=repository;this.audit=audit;this.database=database;this.transactions=transactions;
    }

    public Map<String,Object> companyJob(UUID companyId,UUID jobId){
        Map<String,Object> job=repository.findById(jobId).orElseThrow(this::notFound);
        Map<?,?> company=(Map<?,?>)job.get("company");
        if(!companyId.toString().equals(String.valueOf(company.get("id")))) throw notFound();
        return job;
    }
    public Map<String,Object> recruitingJob(UUID jobId){return repository.findRecruiting(jobId).orElseThrow(this::notFound);}

    public Map<String,Object> createDraft(AuthPrincipal actor,Draft input,RequestMetadata request){
        return transactions.execute(status->{
            String companyStatus=repository.companyStatus(actor.companyId());
            if(companyStatus==null) throw notFound();
            if(!"ACTIVE".equals(companyStatus)) throw new AppException(HttpStatus.FORBIDDEN,"COMPANY_NOT_ACTIVE","Doanh nghiệp chưa ở trạng thái đối tác hoạt động.");
            validateReferences(input);
            UUID id=repository.createDraft(actor.companyId(),actor.userId(),input);
            repository.replaceReferences(id,input);
            repository.addHistory(id,UUID.randomUUID(),null,"DRAFT","COMPANY",actor.userId(),null,null);
            audit.recordAudit(actor.userId(),"JOB_DRAFT_CREATED","JOB_POST",id,Map.of(),request);
            return repository.findById(id).orElseThrow(this::notFound);
        });
    }

    public Map<String,Object> updateDraft(AuthPrincipal actor,UUID jobId,Draft input,RequestMetadata request){
        if(input.expectedVersion()==null) throw new AppException(HttpStatus.BAD_REQUEST,"VALIDATION_ERROR","Dữ liệu gửi lên không hợp lệ.");
        return transactions.execute(status->{
            LockedJob job=repository.lock(jobId).orElseThrow(this::notFound);
            if(!actor.companyId().equals(job.companyId())) throw notFound();
            if(job.status()!=Status.DRAFT&&job.status()!=Status.REVISION_REQUIRED) throw conflict("Chỉ có thể sửa tin ở trạng thái Bản nháp hoặc Cần chỉnh sửa.");
            if(job.version()!=input.expectedVersion()) throw conflict("Tin đã được cập nhật ở nơi khác. Vui lòng tải lại dữ liệu.");
            validateReferences(input);repository.updateDraft(jobId,input);repository.replaceReferences(jobId,input);
            audit.recordAudit(actor.userId(),"JOB_DRAFT_UPDATED","JOB_POST",jobId,Map.of("fromVersion",job.version()),request);
            return repository.findById(jobId).orElseThrow(this::notFound);
        });
    }

    public Map<String,Object> submit(AuthPrincipal actor,UUID jobId,UUID commandId,RequestMetadata request){
        return transactions.execute(status->{
            if(repository.commandExists(jobId,commandId)) return owned(actor.companyId(),jobId);
            LockedJob job=repository.lock(jobId).orElseThrow(this::notFound);
            if(!actor.companyId().equals(job.companyId())) throw notFound();
            if(repository.commandExists(jobId,commandId)) return repository.findById(jobId).orElseThrow(this::notFound);
            if(job.status()!=Status.DRAFT&&job.status()!=Status.REVISION_REQUIRED) throw conflict("Tin không còn ở trạng thái có thể gửi UIT duyệt.");
            if(!"ACTIVE".equals(job.partnerStatus())) throw new AppException(HttpStatus.FORBIDDEN,"COMPANY_NOT_ACTIVE","Doanh nghiệp chưa ở trạng thái đối tác hoạt động.");
            if(job.deadline().isBefore(LocalDate.now(ZoneOffset.UTC))) throw new AppException(HttpStatus.BAD_REQUEST,"JOB_DEADLINE_EXPIRED","Hạn ứng tuyển phải từ hôm nay trở đi.");
            repository.submit(jobId);repository.addHistory(jobId,commandId,job.status().name(),"PENDING_UIT_REVIEW","COMPANY",actor.userId(),null,null);
            Map<String,Object> result=repository.findById(jobId).orElseThrow(this::notFound);
            Map<?,?> company=(Map<?,?>)result.get("company");notifyUit(jobId,commandId,String.valueOf(result.get("title")),String.valueOf(company.get("name")));
            audit.recordAudit(actor.userId(),"JOB_SUBMITTED_FOR_UIT_REVIEW","JOB_POST",jobId,Map.of("fromStatus",job.status().name(),"commandId",commandId.toString()),request);
            return result;
        });
    }

    public Map<String,Object> review(UUID actorId,UUID jobId,UUID commandId,ReviewDecision decision,ReviewReason reason,RequestMetadata request){
        return transactions.execute(status->{
            if(repository.commandExists(jobId,commandId)) return repository.findById(jobId).orElseThrow(this::notFound);
            LockedJob job=repository.lock(jobId).orElseThrow(this::notFound);
            if(repository.commandExists(jobId,commandId)) return repository.findById(jobId).orElseThrow(this::notFound);
            if(job.status()!=Status.PENDING_UIT_REVIEW) throw conflict("Tin đã được xử lý hoặc không còn chờ UIT duyệt.");
            if(decision==ReviewDecision.APPROVE&&job.deadline().isBefore(LocalDate.now(ZoneOffset.UTC)))
                throw new AppException(HttpStatus.BAD_REQUEST,"JOB_DEADLINE_EXPIRED","Không thể duyệt tin đã hết hạn ứng tuyển.");
            Status next=switch(decision){case APPROVE->Status.RECRUITING;case REQUEST_REVISION->Status.REVISION_REQUIRED;case REJECT->Status.REJECTED;};
            repository.review(jobId,next,actorId);
            repository.addHistory(jobId,commandId,"PENDING_UIT_REVIEW",next.name(),"UIT_ADMIN",actorId,
                    reason==null?null:reason.reasonCode().strip(),reason==null?null:reason.note().strip());
            Map<String,Object> result=repository.findById(jobId).orElseThrow(this::notFound);
            Map<?,?> company=(Map<?,?>)result.get("company");notifyCompany(UUID.fromString(String.valueOf(company.get("id"))),jobId,commandId,String.valueOf(result.get("title")),decision,reason);
            Map<String,Object> metadata=new LinkedHashMap<>();metadata.put("commandId",commandId.toString());metadata.put("nextStatus",next.name());metadata.put("reasonCode",reason==null?null:reason.reasonCode());
            audit.recordAudit(actorId,"JOB_UIT_"+decision.name(),"JOB_POST",jobId,metadata,request);
            return result;
        });
    }

    private void validateReferences(Draft input){
        if(!repository.referencesValid("categories",input.categoryIds())||!repository.referencesValid("skills",input.skillIds()))
            throw new AppException(HttpStatus.BAD_REQUEST,"JOB_REFERENCE_INVALID","Nhóm ngành hoặc kỹ năng không hợp lệ.");
    }
    private Map<String,Object> owned(UUID companyId,UUID jobId){return companyJob(companyId,jobId);}

    private void notifyUit(UUID jobId,UUID commandId,String title,String company){
        database.sql("""
          INSERT INTO notifications(recipient_user_id,type,title,body,resource_type,resource_id,deep_link,dedupe_key,payload)
          SELECT u.id,'JOB_PENDING_UIT_REVIEW','Có tin tuyển dụng mới cần duyệt',:company||' đã gửi tin "'||:title||'".','JOB_POST',:jobId,
            '/uit/jobs/review','job:'||CAST(:jobId AS text)||':submit:'||CAST(:commandId AS text)||':'||CAST(u.id AS text),
            jsonb_build_object('jobId',CAST(:jobId AS text),'commandId',CAST(:commandId AS text))
          FROM users u WHERE u.role='UIT_ADMIN' AND u.status='ACTIVE' ON CONFLICT(dedupe_key) DO NOTHING
          """).param("company",company).param("title",title).param("jobId",jobId).param("commandId",commandId).update();
    }
    private void notifyCompany(UUID companyId,UUID jobId,UUID commandId,String jobTitle,ReviewDecision decision,ReviewReason reason){
        String type,title,body;
        switch(decision){
            case APPROVE->{type="JOB_APPROVED";title="Tin tuyển dụng đã được duyệt";body="Tin \""+jobTitle+"\" đã được công khai.";}
            case REQUEST_REVISION->{type="JOB_REVISION_REQUIRED";title="Tin tuyển dụng cần chỉnh sửa";body=reason==null?"Tin cần chỉnh sửa.":reason.note();}
            default->{type="JOB_REJECTED";title="Tin tuyển dụng đã bị từ chối";body=reason==null?"Tin đã bị từ chối.":reason.note();}
        }
        database.sql("""
          INSERT INTO notifications(recipient_user_id,type,title,body,resource_type,resource_id,deep_link,dedupe_key,payload)
          SELECT u.id,:type,:title,:body,'JOB_POST',:jobId,'/company/jobs/'||CAST(:jobId AS text),
            'job:'||CAST(:jobId AS text)||':review:'||CAST(:commandId AS text)||':'||CAST(u.id AS text),
            jsonb_build_object('jobId',CAST(:jobId AS text),'commandId',CAST(:commandId AS text),'decision',:decision)
          FROM company_users cu JOIN users u ON u.id=cu.user_id WHERE cu.company_id=:companyId AND u.status='ACTIVE'
          ON CONFLICT(dedupe_key) DO NOTHING
          """).param("type",type).param("title",title).param("body",body).param("jobId",jobId).param("commandId",commandId)
                .param("decision",decision.name().toLowerCase().replace('_','-')).param("companyId",companyId).update();
    }
    private AppException notFound(){return new AppException(HttpStatus.NOT_FOUND,"JOB_NOT_FOUND","Không tìm thấy tin tuyển dụng.");}
    private AppException conflict(String message){return new AppException(HttpStatus.CONFLICT,"JOB_STATE_CONFLICT",message);}
}
